types/eslint/ts3.1/index.d.ts

```diff
@@ -236,7 +236,13 @@ export namespace Rule {
     }
 
     type NodeTypes = ESTree.Node['type'];
-    type NodeListener = { [T in NodeTypes]?: (node: ESTree.Node) => void };
+    type NodeListener = ESTree.Node extends infer T
+        ? (T extends ESTree.Node ? (_: { [_ in T['type']]?: (node: T) => void }) => void : never) extends (
+              _: infer U,
+          ) => void
+            ? U
+            : never
+        : never;
```

Comment:
I don't think this kind of conditional logic is a good idea
In my experience, creating types dynamically like this over large sets (`ESTree.Node` is a union of 70 interfaces with 68 distinct `type`'s) can easily cause OOMs in TS
https://github.com/microsoft/TypeScript/issues/30473

When I dumped this into a local repo, vscode's typescript integration paused for a few seconds whilst it computed the types.

The only thing that I've found to work efficiently is to list out every single key manually:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/experimental-utils/src/ts-eslint/Rule.ts#L247-L424